File: go/schema/schema.go
package schema

import (
	"encoding/json"
	"fmt"

	"github.com/ter-sdk/ter-go/types"
)

// VariableDefinition represents an environment variable
type VariableDefinition struct {
	Name        string
	Type        types.BaseType
	Default     interface{}
	Required    bool
	Description string
}

// Schema defines environment contract
type Schema struct {
	variables map[string]*VariableDefinition
}

// NewSchema creates a new schema
func NewSchema() *Schema {
	return &Schema{
		variables: make(map[string]*VariableDefinition),
	}
}

// Define adds a variable definition
func (s *Schema) Define(name string, varType types.BaseType) *Schema {
	return s.DefineWithDesc(name, varType, "")
}

// DefineWithDesc adds a variable with description
func (s *Schema) DefineWithDesc(name string, varType types.BaseType, desc string) *Schema {
	s.variables[name] = &VariableDefinition{
		Name:        name,
		Type:        varType,
		Default:     varType.GetDefault(),
		Required:    varType.IsRequired(),
		Description: desc,
	}
	return s
}

// GetVariable returns a variable definition
func (s *Schema) GetVariable(name string) *VariableDefinition {
	return s.variables[name]
}

// GetVariables returns all variables
func (s *Schema) GetVariables() map[string]*VariableDefinition {
	vars := make(map[string]*VariableDefinition)
	for k, v := range s.variables {
		vars[k] = v
	}
	return vars
}

// HasVariable checks if variable exists
func (s *Schema) HasVariable(name string) bool {
	_, ok := s.variables[name]
	return ok
}

// Validate validates values against schema
func (s *Schema) Validate(values map[string]interface{}) (map[string]interface{}, error) {
	errors := make(map[string]string)
	validated := make(map[string]interface{})

	for name, varDef := range s.variables {
		if val, ok := values[name]; ok {
			result := varDef.Type.Validate(val)
			if result.Valid {
				validated[name] = result.Value
			} else {
				errors[name] = result.Error
			}
		} else if varDef.Default != nil {
			validated[name] = varDef.Default
		} else if varDef.Required {
			errors[name] = "Required"
		}
	}

	if len(errors) > 0 {
		return nil, fmt.Errorf("validation errors: %v", errors)
	}

	return validated, nil
}

// ToMap converts schema to map representation
func (s *Schema) ToMap() map[string]interface{} {
	vars := make(map[string]interface{})
	for name, varDef := range s.variables {
		vars[name] = map[string]interface{}{
			"name":        varDef.Name,
			"required":    varDef.Required,
			"default":     varDef.Default,
			"description": varDef.Description,
		}
	}
	return map[string]interface{}{
		"version":   "1.0",
		"variables": vars,
	}
}

// ToJSON converts schema to JSON string
func (s *Schema) ToJSON() (string, error) {
	data := s.ToMap()
	b, err := json.MarshalIndent(data, "", "  ")
	if err != nil {
		return "", err
	}
	return string(b), nil
}

File: src/core/schema.ts
/**
 * Schema Definition for TER
 * Defines environment contracts with typed variables
 */

import { BaseType, TypeValidationError } from './types';

export interface VariableDefinition {
  name: string;
  type: BaseType;
  isRequired: boolean;
  defaultValue?: unknown;
  description?: string;
  policy?: {
    readonly?: boolean;
    visibility?: 'public' | 'secret' | 'internal';
  };
}

export interface SchemaValidationResult {
  valid: boolean;
  errors: Array<{
    variable: string;
    error: TypeValidationError;
  }>;
}

/**
 * Environment Contract Schema
 */
export class Schema {
  private variables = new Map<string, VariableDefinition>();

  /**
   * Define a variable in the schema
   */
  define(name: string, type: BaseType): this {
    if (this.variables.has(name)) {
      throw new Error(`Variable "${name}" is already defined`);
    }

    this.variables.set(name, {
      name,
      type,
      isRequired: type.isRequired,
      defaultValue: type.defaultValue,
    });

    return this;
  }

  /**
   * Get a variable definition by name
   */
  getVariable(name: string): VariableDefinition | undefined {
    return this.variables.get(name);
  }

  /**
   * Get all variables
   */
  getVariables(): VariableDefinition[] {
    return Array.from(this.variables.values());
  }

  /**
   * Check if a variable is defined
   */
  hasVariable(name: string): boolean {
    return this.variables.has(name);
  }

  /**
   * Validate values against the schema
   */
  validate(values: Record<string, string | undefined>): SchemaValidationResult {
    const errors: Array<{
      variable: string;
      error: TypeValidationError;
    }> = [];

    for (const [name, definition] of this.variables) {
      const value = values[name];
      const result = definition.type.validate(value);

      if (!result.valid) {
        const error = result.error;
        error.field = name;
        errors.push({
          variable: name,
          error,
        });
      }
    }

    return {
      valid: errors.length === 0,
      errors,
    };
  }

  /**
   * Get schema as JSON for inspection
   */
  toJSON() {
    return {
      version: '1.0',
      variables: Array.from(this.variables.values()).map((v) => {
        const varJSON: any = {
          name: v.name,
          type: v.type.name,
          required: v.type.isRequired,
          defaultValue: v.defaultValue,
          description: v.description,
          isSecret: v.type.isSecret,
          policy: v.policy,
        };
        
        // Store enum values if applicable
        if (v.type.name === 'enum' && (v.type as any).allowedValues) {
          varJSON.values = (v.type as any).allowedValues;
        }
        
        return varJSON;
      }),
    };
  }
}

/**
 * Schema builder for fluent API
 */
export class SchemaBuilder {
  private schema = new Schema();

  /**
   * Define a variable with fluent API
   */
  var(name: string, type: BaseType): this {
    this.schema.define(name, type);
    return this;
  }

  /**
   * Build and return the schema
   */
  build(): Schema {
    return this.schema;
  }
}

/**
 * Helper to create a schema
 */
export function defineSchema(): SchemaBuilder {
  return new SchemaBuilder();
}

/**
 * Helper to create a schema from object
 */
export function createSchema(definitions: Record<string, BaseType>): Schema {
  const schema = new Schema();
  for (const [name, type] of Object.entries(definitions)) {
    schema.define(name, type);
  }
  return schema;
}
